src/net/atlas/download.rs

```diff
@@ -0,0 +1,1025 @@
+                                Some(report) => report.clone(),
+                                None => ReliabilityReport::empty(),
+                            };
+                            peers.insert(peer_url, report);
+                        }
+                    }
+                    let ctx = AttachmentsBatchStateContext::new(
+                        attachments_batch,
+                        peers,
+                        &network.connection_opts,
+                    );
+                    AttachmentsBatchStateMachine::new(ctx)
+                }
+                None => {
+                    // Nothing to do!
+                    return Ok(vec![]);
+                }
+            },
+        };
+
+        let mut progress =
+            AttachmentsBatchStateMachine::try_proceed(ongoing_fsm, dns_client, network, chainstate);
+
+        match progress {
+            AttachmentsBatchStateMachine::Done(ref mut context) => {
+                for attachment in context.attachments.drain() {
+                    let mut attachments_instances = network
+                        .atlasdb
+                        .find_all_attachment_instances(&attachment.hash)
+                        .map_err(|e| net_error::DBError(e))?;
+                    network
+                        .atlasdb
+                        .insert_new_attachment(&attachment.hash, &attachment.content, true)
+                        .map_err(|e| net_error::DBError(e))?;
+                    resolved_attachments.append(&mut attachments_instances);
+                    context
+                        .attachments_batch
+                        .resolve_attachment(&attachment.hash)
+                }
+
+                // Update reliability reports
+                for (peer_url, report) in context.peers.drain() {
+                    self.reliability_reports.insert(peer_url, report);
+                }
+
+                // Re-insert AttachmentsBatch back to the queue if not fully processed
+                if !context.attachments_batch.has_fully_succeed() {
+                    context.attachments_batch.bump_retry_count();
+                    self.priority_queue.push(context.attachments_batch.clone());
+                }
+            }
+            next_state => {
+                self.ongoing_batch = Some(next_state);
+            }
+        };
+
+        Ok(resolved_attachments)
+    }
+
+    pub fn enqueue_new_attachments(
+        &mut self,
+        new_attachments: &mut HashSet<AttachmentInstance>,
+        atlasdb: &mut AtlasDB,
+    ) -> Result<Vec<AttachmentInstance>, net_error> {
+        if new_attachments.is_empty() {
+            return Ok(vec![]);
+        }
+
+        let mut resolved_attachments = vec![];
+        let mut attachments_batch = AttachmentsBatch::new();
+        for attachment in new_attachments.drain() {
+            // Are we dealing with an empty hash - allowed for undoing onchain binding
```

Comment:
I'm not sure what you mean in this comment?

Reply:
Basically this:
```
(print {
    attachment: {
        hash:0x,
        page-index: page,
        position-in-page: index,
        metadata: {
            name: name,
            namespace: namespace
        }
    }})
```

Reply:
Ah, I see.